### PR TITLE
Bugfix ktps 1929 train val split

### PR DESCRIPTION
**Pull request by FrankKr**
Improves stratification of train and validation set.
Before: [min_peaks+max_peaks]  are equally divided, now they are divided separately.

Additionally, I removed two returns of the function, which were only used for testing.

### CODE REVIEW

File: openstef/model_selection/model_selection.py [hunk 1]
```diff
@@ -175,54 +174,47 @@ def split_data_train_validation_test(
...
-
-    if peak_n_days < 3:
-        stratification_min_max = (
-            False  # stratification is not adding value in this case
+    if stratification_min_max and (len(set(train_val_data.index.date)) >= 4):
+        # First determine the dates with min and max values
+        # Let's consider the 15% lowest or highest days, with a minimum
+        # of two dates in the min and max list
+        train_val_dates = list(set(train_val_data.index.date))
+        n_days_per_subset = int(max(0.15 * len(set(data_.index.date)), 2))
```
[the remark this answers is not in the document]
Comment (JanMaartenvanDoorn) [reply openstef/model_selection/model_selection.py:182]: maybe the same for the minium of two days

File: openstef/model_selection/model_selection.py [hunk 2]
```diff
@@ -175,54 +174,47 @@ def split_data_train_validation_test(
...
-        split_val = int((peak_n_days * validation_fraction) / PERIOD_TIMEDELTA)
-        # Always have at least one validation split
-        split_val = max(split_val, 1)
-        peaks_val, idx_val_split = sample_indices_train_val(
-            data_,
-            random_sample(np.array(min_max_dates.index.date), k=split_val),
+        # Find min_dates, but do not consider the max_dates
+        min_dates_subset = train_val_data.loc[
+            ~np.isin(train_val_data.index.date, max_dates), ["load"]
+        ]
```
Comment (JanMaartenvanDoorn) [inline openstef/model_selection/model_selection.py:195]: clever

File: test/unit/model/test_model_selection.py [hunk 3]
```diff
@@ -79,22 +78,67 @@ def test_split_data_train_validation_test_stratification(self):
...
             len(data.index) * SPLIT_PARAMS["test_fraction"],
             delta=4,
         )
 
+    def test_split_data_train_validation_test_stratification(self):
+        """
+        Test that the train/validation split is stratified,
+        meaning min/max days are equally distributed in train and validation set
+        """
+        df = pd.DataFrame(
```
Comment (JanMaartenvanDoorn) [inline test/unit/model/test_model_selection.py:97]: # Arrange

File: test/unit/model/test_model_selection.py [hunk 4]
```diff
@@ -79,22 +78,67 @@ def test_split_data_train_validation_test_stratification(self):
...
+        min_days = random.sample(
+            [x for x in set(df.index.day) if x not in max_days], n_days_low
+        )
+        # Either increase or decrease load on those days
+        for day in max_days:
+            df.loc[df.index.day == day, "load"] += 5
+        for day in min_days:
+            df.loc[df.index.day == day, "load"] -= 5
+
+        # Split using default arguments. Should result in stratified split
```
Comment (JanMaartenvanDoorn) [inline test/unit/model/test_model_selection.py:121]: Act

File: test/unit/model/test_model_selection.py [hunk 5]
```diff
@@ -79,22 +78,67 @@ def test_split_data_train_validation_test_stratification(self):
...
+        (train, val, test,) = model_selection.split_data_train_validation_test(
+            df, test_fraction=0, stratification_min_max=True
+        )
+        # test that max and min days are both in train and val sets
+        for dayset in [max_days, min_days]:
```
Comment (JanMaartenvanDoorn) [inline test/unit/model/test_model_selection.py:126]: Assert

File: test/unit/model/test_model_selection.py [hunk 6]
```diff
@@ -79,22 +78,67 @@ def test_split_data_train_validation_test_stratification(self):
...
+            delta=2 * 96,
+        )  # two days is allowed
 
         self.assertAlmostEqual(
             len(test_set),
             len(data.index) * SPLIT_PARAMS["test_fraction"],
             delta=4,
         )
 
+    def test_split_data_train_validation_test_stratification(self):
```
Comment (JanMaartenvanDoorn) [inline test/unit/model/test_model_selection.py:92]: Nice unit test, suggestion: May document in the following way: https://wiki.c2.com/?ArrangeActAssert

File: test/unit/model/test_model_selection.py [hunk 7]
```diff
@@ -79,22 +78,67 @@ def test_split_data_train_validation_test_stratification(self):
...
+                if n_days_in == 0:
+                    print("Investigate this!")
+
+                assert n_days_in >= 1
+
+    def test_split_data_train_validation_test_stratification_repeat_20_times(self):
+        """Repeat the stratification test 20 times,
+        since the test using random selection and should work every time"""
+        for _ in range(20):
+            self.test_split_data_train_validation_test_stratification()
```
Comment (JanMaartenvanDoorn) [inline test/unit/model/test_model_selection.py:140]: Clever way of testing random :)

**Review by JanMaartenvanDoorn — COMMENTED** [review]
Nice improvement! I added some suggestions, I think the  sonarcloud warnin should be addressed.

File: openstef/model_selection/model_selection.py [hunk 8]
```diff
@@ -175,54 +174,47 @@ def split_data_train_validation_test(
...
+        other_dates = [
+            x for x in train_val_dates if x not in min_dates and x not in max_dates
+        ]
+
+        # Divide min, max and other dates fairly over validation and train set, with at least 1 min and max in train and validation
+        val_dates = []
+        train_dates = []
+        for date_set in [max_dates, min_dates, other_dates]:
+            n_days_val = max(1, int(validation_fraction * len(date_set)))
+            val_dates += random.sample(list(date_set), n_days_val)
```
[the remark this answers is not in the document]
Comment (JanMaartenvanDoorn) [reply openstef/model_selection/model_selection.py:213]: Maybe picking the numpy variant will solve the problem

File: test/unit/model/test_model_selection.py (hunk 6 above)
Comment (FrankKr) [reply test/unit/model/test_model_selection.py:92]: Good idea

File: test/unit/pipeline/test_train.py [hunk 9]
```diff
@@ -56,15 +51,14 @@ def test_split_data_train_validation_test_stratification_no_backtest(self):
         # delta = 1, number of the peaks the two amounts may differ for the train and validation data
         # delta = 4, when looking at the test data, can differ 1 hr (4x15min)
 
+        # Add checkpoint, useful for debugging when things go wrong
```
Comment (JanMaartenvanDoorn) [inline test/unit/pipeline/test_train.py:54]: Commented out code

**Review by JanMaartenvanDoorn — APPROVED** [review]

**Pull request merged**